### ISSUE TITLE
Direct support for conditional single-instance resources

### ISSUE THREAD

**Issue by staizen-stephen** [opentofu/opentofu#1306]
### OpenTofu Version

```shell
None. Currently using Hashicorp Terraform.
```

### Use Cases

We often use `count = var.xxxx ? 1 : 0` in various flavours but with the same purpose of determining to create or not create a single resource.

While this is currently supported it means that resources require an indexed reference, and this also means that if you later change something to be conditional then you have a lot of manual state manipulation required.

I'd like to see this natively supported as it would simplify a lot of code.

### Attempted Solutions

Workarounds are using `count = ...` expressions.

### Proposal

Add an additional meta-data atttribute `opentofu-options` and a sub-attribute `resource-enabled`, for example:

```
resource "aws_key_pair" "my-key-pair" {
  ...
  opentofu-options {
    resource-enabled = var.create-key-pair
  }
}
```

When a resource is not enabled (known at plan time), then any reference to it in the graph becomes invalid and this would prevent a bunch of munging such as `coalesce(aws-key_pair.my-key-pair.*.name, "")` or equivalent.

### References

* https://github.com/hashicorp/terraform/issues/21953 - long thread, lots of motivation, no movement

**Comment by cam72cam** [comment]
I've marked this as needs-rfc as it requires some pretty careful consideration around how this could be designed.  https://github.com/hashicorp/terraform/issues/21953#issuecomment-1433464018 is a good summary of the current state of discussion.

I see this as two distinct discussions that are linked.  One is to add the enabled flag to resources/modules/etc... and handle the result of type.name.attr being an invalid reference if disabled.  The other is syntax sugar around accessing these optional values.

On initial reading, I'm leaning toward setting a disabled resource/module/etc... to a typed unknown and making sure that references behave as expected, using expressions like `type.name =! null ? type.name.attr : other_value`.

I'd recommend that any discussion of "syntax sugar" like the `if_enabled` or `try` functions happen in a followup issue.

**Comment by cam72cam** [comment]
I think part of what was not explained well in the original issue is that the evaluator needs to know all of the types / existence in an expression in order for it to be valid.

In the example:

```hcl
resource "foo" "bar" {
   enabled = local.isenabled
}

output "myout" {
   value = local.isenabled ? foo.bar.attribute : null
}
```

In that example, we need to know that foo.bar.attribute could exist and is of a certain type.  HCL does not do any short circuiting currently, even for && and || operators.  These sort of implementation details muddied the waters of the original issue.  We need to remember that HCL is a declarative config language that is being shoe-horned into being something close to a programming language.

**Comment by cam72cam** [comment]
I think it's as simple as making foo.bar a "typed unknown" or null value.  We know that foo.bar could be a "foo resource" or it might not exist.  We don't care if it exists or not until we actually try to access the value.  Knowing that it might have a `attribute` should be enough for the expression traversal not to complain, though that level of knowledge may not even be nessesary.

Again, this is getting into the weeds of the OpenTofu and HCL internals and might be confusing the issue.

From a user facing perspective, `local.isenabled ? foo.bar.attribute : null` would work as intended.  I also believe it would be fairly easy to support `foo.bar != null ? foo.bar.attribute : null`.  If you messed up or forgot the ternary expression and tried to access `foo.bar.attribute` when it is not enabled it would throw an error.

Would allowing those two forms of expression suffice for your use cases?

I also think you could get fancy with it, without adding syntax sugar:
```hcl
resource "foo" "bar" {
  enabled = var.isenabled
}

locals {
  fallback = {attr = "default_value")
}

output "out" {
// Four Equivalent expressions

  // 1. Use the same expression as the enabled field
  value = var.isenabled ? foo.bar.attr : "default_value"
  
  // 2. Check if the resource is null directly
  value = foo.bar != null ? foo.bar.attr : "default_value"
  
  // 3. Use coalesce with an inline default
  value = coalesce(foo.bar, {attr = "default_value"}).attr
  
  // 4. Use coalesce with a local default 
  value = coalesce(foo.bar, local.fallback).attr
}
```

All four of those would work under the same set of logic I'm proposing above.  I'd probably only ever use 2 or 4 myself, but it is flexible to do any of those four variants.

**Comment by serialseb** [comment]
Retranscripting my chat from github:

The typical example is this:
```
variable "generate_policy" { type = bool }
resource "aws_iam_policy" "myPolicy" {
  if = var.generate_policy
}
```
In replacement of the common occurence i see which makes a [0] a part of weird code

```
variable "generate_policy" { type = bool }
resource "aws_iam_policy" "myPolicy" {
  count = var.generate_policy == true ? 1 : 0
}
```

it can then flow naturally considering the same syntax as depends_on, returning a bool in those circumstances it's being tested against:

```
resource "aws_iam_policy_role_attachment" "aws_managed" {
  if = aws_iam_poliy.myPolicy
  policy_arn = aws_iam_policy.myPolicy
  role = "MyDefinedRole"
}
```

I think it's a choice between introduing null or making the resource booley in those scenarios (null not being my favourite keyword in any language)

**Comment by tom-reinders** [comment]
Would this also be considered for dynamic blocks?

```terraform
resource "azurerm_linux_function_app" "function_app" {
  // ...

  site_config {
    // ...

    dynamic "application_stack" {
      for_each = var.dotnet_version != "" ? [var.dotnet_version] : []
      content {
        dotnet_version              = application_stack.value
        use_dotnet_isolated_runtime = var.dotnet_isolated
      }
    }

    // ...
  }

  // ...
}
```

becomes something like

```terraform
resource "azurerm_linux_function_app" "function_app" {
  // ...

  site_config {
    // ...

    dynamic "application_stack" {
      if = var.dotnet_version != ""
      content {
        dotnet_version              = var.dotnet_version
        use_dotnet_isolated_runtime = var.dotnet_isolated
      }
    }

    // ...
  }

  // ...
}
```

**Comment by cam72cam** [comment]
@tom-reinders I think it's a different set of technical issues, but should follow the same syntax.

**Comment by ikegentz** [comment]
Don't want to distract from the main discussion but wanted to throw this out there; as a TF user that's been tracking https://github.com/hashicorp/terraform/issues/21953 for some time (and the [previous issue](https://github.com/hashicorp/terraform/issues/30221) dating back to 2021), it appears there's a lot of things to think about from a backwards-compatibility perspective. FWIW, I'd happily refactor my code to a new major version if that's what it takes to get this over the line, instead of trying to design it to be perfectly backwards-compatible with all `1.y.z` versions

**Comment by RockyMM** [comment]
Speaking as a member of the community:

@ikegentz 
`enabled` could be a feature of `1.9+.y.z` version, not necessarily `2.y.z`.

But I just noticed this issue. Is `opentofu-options` a thing? I find the usage of the kebab case very un-Terraformy. 

However, probably in the future something like `_opentofu_options` should be a thing. Is there an issue where this is discussed?

**Comment by staizen-stephen** [comment]
> Speaking as a member of the community:
> 
> @ikegentz `enabled` could be a feature of `1.9+.y.z` version, not necessarily `2.y.z`.
> 
> But I just noticed this issue. Is `opentofu-options` a thing? I find the usage of the kebab case very un-Terraformy.
> 
> However, probably in the future something like `_opentofu_options` should be a thing. Is there an issue where this is discussed?

Sorry, my fault. `opentofu-options` was my invention for the discussion. It should obviously match the style of the rest of the platform, that also includes other fields.

**Comment by apparentlymart** [comment]
One specific design detail to think about here is whether OpenTofu can and should be able to give immediate feedback to someone who retrofits `enabled` without making sure that all of the existing references to the resource are now compatible with the resource object being null.

With `count` today, the change of type from object to tuple-of-object effectively forces all of the references to immediately fail type-checking, prompting the author to update them all to properly support the case where there are zero instances.

For example, let's start with a simple module that doesn't have any dynamic enablement supported at all:

```hcl
resource "null_resource" "example" {
}

output "thingy_id" {
  value = null_resource.example.id
}
```

If I later decide to make that resource dynamically-enableable using a boolean input variable and `count`, but I forget to update the `output "thingy_id"` block, then OpenTofu catches my mistake immediately due to the type mismatch:

```hcl
variable "thingy_enabled" {
  type    = bool
  default = true
}

resource "null_resource" "example" {
  count = var.thingy_enabled ? 1 : 0
}

output "thingy_id" {
  value = null_resource.example.id
}
```

```
╷
│ Error: Missing resource instance key
│ 
│   on example.tf line 12, in output "thingy_id":
│   12:   value = null_resource.example.id
│ 
│ Because null_resource.example has "count" set, its attributes must be
│ accessed on specific instances.
│ 
│ For example, to correlate with indices of a referring resource, use:
│     null_resource.example[count.index]
╵
```

This error message is currently admittedly not quite right for an `output` block since an output value doesn't have its own instances to "correlate with" the other resource, but the typical fix here would be to change this to something like...

```hcl
output "thingy_id" {
  value = one(null_resource.example[*].id)
}
```

...which means that the `thingy_id` output would be set to `null` if `var.thingy_enabled` is false, or be a non-null string if it's true.

---

If we instead imagine the hypothetical of supporting an `enabled` argument which makes `null_resource.example` appear `null` when it's set to `false`, then this isn't so obvious...

```hcl
variable "thingy_enabled" {
  type    = bool
  default = true
}

resource "null_resource" "example" {
  enabled = var.thingy_enabled
}

output "thingy_id" {
  value = null_resource.example.id
}
```

Since `var.thingy_enabled` defaults to `true`, `null_resource.example` will be a non-null object unless I specifically remember to test what happens when that variable is set to false. Unfortunately the OpenTofu language doesn't track nullability as part of the type system, there's no way for the validator to notice using only static analysis that a bare `null_resource.example.id` is no longer valid and must instead be rewritten like this in order to work properly:

```hcl
output "thingy_id" {
  value = one(null_resource.example[*].id)
}
```

or, if you prefer, equivalently:

```hcl
output "thingy_id" {
  value = null_resource.example != null ? null_resource.example.id : null
}
```

Of course, with this short and contrived configuration it's easy to find and fix the problem, but in larger modules with many different overlapping dynamic options it's considerably harder to exhaustively test all of the combinations and so the current design of `count` is providing some static type checking as a side-effect, even though that's admittedly an accidental benefit that was formalized in 0.12 rather than being intentionally designed that way from the outset.

---

I expect different folks will have a different sense of how much this specific problem matters.

In particular, those who are accustomed to languages with totally-dynamic type systems are presumably already accustomed to relying on large suites of unit tests instead of on compile-time checks made by a compiler, and so I expect that cohort would find the above situation pretty consistent with their expectations coming from other languages.

So far OpenTofu language design has tried to lean toward catching as much as possible during static validation -- the closest idea we have to "compile-time" -- so that module authors can focus their unit testing efforts primarily on the unique behaviors of their module, rather than on low-level concerns like type/value correctness. Implementing this as I've described would therefore be a regression of that goal. That doesn't mean we _definitely shouldn't_ do it, but it does seem like a prompt to pause and consider the tradeoffs.

(In an ideal world I'd like to support `enabled` while still finding some way to static-check it, but I don't have a concrete idea about how to do that right now without having a time machine to go back and introduce explicitly-nullable types in an earlier version.)

**Comment by apparentlymart** [comment]
It's also worth noting that adding any new "meta-arguments" to `resource`/`data`/`module` blocks is technically a breaking change, because the meta-arguments effectively shadow the namespace of valid resource argument names or module input variable names.

I have definitely seen real modules with `variable "enabled"` declarations, so this is not just a hypothetical.

There is some special escaping syntax included as an escape hatch for situations like this, where arguments placed in a block literally named `_` (just an underscore) are forced to be treated as non-meta-arguments:

```hcl
module "example" {
  _ {
    enabled = true
  }
}
```

...but that mechanism was intended to be used in conjunction with introducing a new [language edition](https://log.martinatkins.me/2021/09/25/future-of-the-terraform-language/#terraform-language-editions), so that a hypothetical automatic edition upgrade tool could automatically rewrite a module to specify a new edition in the `language` argument and rewrite any conflicting arguments like the above to be inside such an escaping block, while existing modules that don't opt in to the new edition would not be able to use `enabled` so that they don't get broken by this addition.

I don't know if we want to go so far as introducing a new language edition to get this new feature, but I mention it just in case it's useful to someone working on an RFC for this in future.

**Comment by apparentlymart** [comment]
I keep coming back to think about this every now and then and have now considered a variety of different weird workarounds for the backward-compatibility concern I described in my previous comment.

Although none of them are _awesome_ I think on reflection my favorite is the following:

To start, we'd introduce `enabled` as an argument inside the `lifecycle` block instead of at the top-level, like this:

```hcl
resource "aws_instance" "example" {
  # ...

  lifecycle {
    enabled = var.enable_server
  }
}
```

The content of `lifecycle` is entirely under OpenTofu's control, so there is no risk of conflict with provider schema or module interfaces there.

Then at some future point -- perhaps once we've accumulated a number of these odd tricks for backward compatibility, including potentially uses of https://github.com/opentofu/opentofu/pull/2262 -- we can introduce a new language edition that would, for modules that opt in to it, move `enabled` up to be declared in the same position as `count`/`for_each`, with modules using that edition then required to use the "escaping block" syntax if they want to call a module or resource type where `enabled` is in the schema. Both of those rewrites could be applied automatically by a hypothetical [edition upgrade tool](https://log.martinatkins.me/2021/09/25/future-of-the-terraform-language/#automatic-edition-migrations), once the time comes.

This compromise allows us to delay the language edition question at the expense of some slightly-more-awkward syntax in the meantime. Aside from that syntax difference I'd imagine the functionality being essentially the same as what I prototyped in https://github.com/opentofu/opentofu/pull/2299.

**Comment by diofeher** [comment]
Hey, OpenTofu community. I've opened a RFC related to this issue and would like to hear your feedback at https://github.com/opentofu/opentofu/pull/3066.

**Comment by diofeher** [comment]
I'm closing this issue since `enabled` is already on `main`, but the work for it is not yet finished. The rest of the work is being tracked here: https://github.com/opentofu/opentofu/issues/3247
